src/test/java/api/requests/RequestsAPICreationTests.java

```diff
@@ -360,9 +363,13 @@ public void cannotCreateRecallRequestWhenItemIsNotCheckedOut() {
     Response postResponse = requestsClient.attemptCreate(new RequestBuilder()
       .recall()
       .withItemId(itemId)
+      .withPickupServicePointId(servicePointsFixture.cd1().getId())
       .withRequesterId(usersFixture.charlotte().getId()));
 
     assertThat(postResponse, hasStatus(HTTP_UNPROCESSABLE_ENTITY));
+    assertThat(postResponse.getJson(), hasErrors(1));
```

Comment:
Some of the validations depend on each other. One of the main challenges of this work was to perform as many validations as possible, while avoiding checks that are guaranteed to fail if any upstream validation in their corresponding dependency chains fail. For example: if we couldn't fetch an item, all the subsequent validations that require an item object would also fail, without providing any additional useful information to the caller other than "an NPE was caught while validating X since item is null". I tried to avoid performing those redundant checks to keep the response concise, informative and actionable.

Checking the number of returned errors in tests verifies that no redundant validations were made.

Reply:
> Some of the validations depend on each other. One of the main challenges of this work was to perform as many validations as possible, while avoiding checks that are guaranteed to fail if any upstream validation in their corresponding dependency chains fail. For example: if we couldn't fetch an item, all the subsequent validations that require an item object would also fail, without providing any additional useful information to the caller other than "an NPE was caught while validating X since item is null". I tried to avoid performing those redundant checks to keep the response concise, informative and actionable.

How does the system avoid performing those checks, given that the initial error is no longer considered fatal and the information about the failure is not available to the validators that occur later?

Reply:
Some existing validations have built-in protection against invalid arguments, so no additional changes were needed. For example: validation that checks if the user is inactive does nothing if the user is null.

For validations that don't have such protection, I used a different approach. While, as you rightly pointed out, the information about the failure is not available to the validators, it is available to the code that calls those validators. So we can check if it makes sense to run a specific validation based on the set of errors accumulated within error handler. See `CreateErrorRequestService#createRequest` for an example:
```
...
.thenComposeAsync(r -> r.after(when(this::shouldCheckItem, this::checkItem, this::doNothing)))
...
```
```
private CompletableFuture<Result<Boolean>> shouldCheckItem(RequestAndRelatedRecords records) {
  return ofAsync(() -> errorHandler.hasNone(INVALID_ITEM_ID));
}
```

Reply:
> For validations that don't have such protection, I used a different approach. While, as you rightly pointed out, the information about the failure is not available to the validators, it is available to the code that calls those validators. 

Because this state is kept by the instance of the service?

So we can check if it makes sense to run a specific validation based on the set of errors accumulated within error handler. 

Is it reasonable me to me to summarise this as managing the flow of the process (business logic) with the presence of an error (and so the error production and these checks are coupled together)?

If we take that example, if an item isn't found, then significant parts of the business logic no longer make sense (e.g. circulation rules and policies) and the client must correct it to make any progress. What is the goal of prolonging the flow, to get a few more errors reported in the first pass?

Reply:
> If we take that example, if an item isn't found, then significant parts of the business logic no longer make sense (e.g. circulation rules and policies) and the client must correct it to make any progress. What is the goal of prolonging the flow, to get a few more errors reported in the first pass?

Yes, the client must correct this issue to continue, but if there are other issues that client can't or doesn't want to correct, he or she may decide that it's not worth it to try to correct the first issue. In this example (when the item is missing) let's say the patron is also blocked from requesting. The librarian may decide to not spend any time finding out what happened to that item because the request will be refused anyway. That was the initial idea - collect as many issues as we possibly can and present all of them to the end user. With these PRs we actually don't go as far, we're just doing this in case when overridable blocks exist to save users from the hassle of overriding these blocks. In other scenarios we just fail with the first error to not break the existing behavior. But later on, with additional work on BE and FE side, we would be able to do this in other scenarios, because all of the errors are already collected. Well, maybe not all of them, just those we can collect during the "refuse when" stage of the process.

Reply:
@alexanderkurash 

> In this example (when the item is missing) let's say the patron is also blocked from requesting. The librarian may decide to not spend any time finding out what happened to that item because the request will be refused anyway.

If the item being not found is likely to cause the client to abort the check out process entirely. What is the value of accumulating other errors with it?

> With these PRs we actually don't go as far, we're just doing this in case when overridable blocks exist to save users from the hassle of overriding these blocks. In other scenarios we just fail with the first error to not break the existing behavior.

Hmm, I took a different understanding from @OleksandrVidinieiev , that errors like item not found do not stop the whole process, instead the other errors are accumulated, however they are thrown away at the end before reporting to the client.

Have I misunderstood this?

Reply:
> If the item being not found is likely to cause the client to abort the check out process entirely. What is the value of accumulating other errors with it?

Well, I doubt that the patron and the librarian will just accept it and forget about their attempt to check out this item. It's likely that they'll try to find another item or figure out what happened to this item (e.g. why it was deleted and by whom). If they knew from the beginning that it won't be possible to do what they're doing anyway (because of item limit block, for example), it would save them some time. Or at least the library can deal with this item in a "lazy" fashion, without a patron breathing down their neck. Of course, this is just my understanding, we probably need POs to comment on such business-related matters.

> Hmm, I took a different understanding from @OleksandrVidinieiev , that errors like item not found do not stop the whole process, instead the other errors are accumulated, however they are thrown away at the end before reporting to the client.

Sorry, when I said "we just fail with the first error to not break the existing behavior" I didn't mean that we fail with it immediately. I meant that we accumulate them and then fail with the first of the accumulated errors, emulating the existing behavior. In the future we can change this behavior to return all errors in all other scenarios (non-override), but it means a lot of work on both BE and FE sides, and it's not directly related to overrides so we've skipped it at this point.

Reply:
> Well, I doubt that the patron and the librarian will just accept it and forget about their attempt to check out this item. It's likely that they'll try to find another item or figure out what happened to this item (e.g. why it was deleted and by whom). 

Sure. I might propose that changing either the item or patron for a check out is a fundamentally different instance of the process, and as such, much of the information no longer relevant.

> If they knew from the beginning that it won't be possible to do what they're doing anyway (because of item limit block, for example), it would save them some time. 

Sure, it would. An item block is an interesting example of this, it relies on both the item and patron to be correct and valid. If either are not, then any pre-emptive checks on item blocks are redundant.

> Or at least the library can deal with this item in a "lazy" fashion, without a patron breathing down their neck. Of course, this is just my understanding, we probably need POs to comment on such business-related matters.

That would be a good example for a PO to advise on how the process works in practice.

> Sorry, when I said "we just fail with the first error to not break the existing behavior" I didn't mean that we fail with it immediately. I meant that we accumulate them and then fail with the first of the accumulated errors, emulating the existing behavior. In the future we can change this behavior to return all errors in all other scenarios (non-override), but it means a lot of work on both BE and FE sides, and it's not directly related to overrides so we've skipped it at this point.

I think this is what I don't understand, why accumulate the non-fatal errors if they are discarded before the response is issued. At which point, they provide no value to the client or user of the system.

Are you saying that this complexity is to facilitate a possible future change in behaviour?

Reply:
> I think this is what I don't understand, why accumulate the non-fatal errors if they are discarded before the response is issued. At which point, they provide no value to the client or user of the system.
> 
> Are you saying that this complexity is to facilitate a possible future change in behaviour?

Yes. Currently we need to accumulate errors only for our new overrides, but this also creates a base for a (possible) future change in behavior in all other scenarios.

Do you think it would be better to just collect validation errors and fail immediately on fatal errors? I think this would also work for our new override logic.

BTW another PR is coming, similar to this one, with the changes in the check-out process. So I'm trying to figure out the correct approach to minimize such discussions later.

Reply:
> Yes. Currently we need to accumulate errors only for our new overrides, but this also creates a base for a (possible) future change in behavior in all other scenarios.
> 
> Do you think it would be better to just collect validation errors and fail immediately on fatal errors? I think this would also work for our new override logic.

I think that satisfies the current required behaviour and likely reduces the complexity in how the code turns a fatal error into a non-fatal error and back again throughout the process.
 
> BTW another PR is coming, similar to this one, with the changes in the check-out process. So I'm trying to figure out the correct approach to minimize such discussions later.

Sure, I appreciate that. I only provided feedback on this design by request and wasn't expecting any substantial changes to it at this point.